In [1]:
import pandas as pd
import pandas as pd
from sqlalchemy import create_engine
%load_ext sql

In [2]:
engine=create_engine('postgresql+psycopg2://postgres:password@localhost:5432/medika_sejahtera')
connection=engine.connect()
print("Connected:", connection.closed == False)

Connected: True


In [3]:
%sql postgresql+psycopg2://postgres:password@localhost:5432/medika_sejahtera

'Connected: postgres@medika_sejahtera'

# 1. Product & Reason Analysis

In [4]:
%%sql

SELECT
    return_reason_cleaned,
    COUNT (*) AS total_record
FROM
    return_cleaned
GROUP BY
    return_reason_cleaned
ORDER BY
    total_record DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


return_reason_cleaned,total_record
Salah Input Sistem,28
Salah Pesan Customer,20
Barang Rusak,7
Kadaluarsa,4


In [14]:
%%sql
WITH total AS
(
    SELECT
        COUNT(*) AS total_records
    FROM
        return_cleaned
)

SELECT
    rc.return_reason_cleaned,
    COUNT(*) AS total_record,
    tl.total_records AS total_data,
    ROUND((COUNT(*)::numeric / NULLIF(tl.total_records,0)::numeric) * 100,2) AS rate
FROM
    return_cleaned rc
CROSS JOIN
    total tl
GROUP BY
    rc.return_reason_cleaned,
    total_data
ORDER BY
    total_record DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


return_reason_cleaned,total_record,total_data,rate
Salah Input Sistem,28,59,47.46
Salah Pesan Customer,20,59,33.90
Barang Rusak,7,59,11.86
Kadaluarsa,4,59,6.78


**Findings**  
Salah Input Sistem (system input error) was the dominant return reason in 2024, accounting for 28 of 59 cases (47.46%). It led Salah Pesan Customer (customer ordering error) — 20 cases (33.90%) — by a notable 13.56 percentage points. Together, these two reasons made up over 81% of all return cases, while Barang Rusak (damaged goods, 11.86%) and Kadaluarsa (expired, 6.78%) contributed far less.

In [ ]:
%%sql

SELECT
    CASE
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3 THEN 'Q1'
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6 THEN 'Q2'
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9 THEN 'Q3'
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12 THEN 'Q4'
    END AS quarter,
    COUNT(*) AS total_record
FROM 
    return_cleaned
WHERE 
    return_reason_cleaned = 'Salah Input Sistem'
GROUP BY 
    quarter
ORDER BY 
    quarter

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


quarter,total_record
Q1,2
Q2,10
Q3,6
Q4,10


In [19]:
%%sql

WITH qr AS (
    SELECT
    CASE
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3 THEN 'Q1'
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6 THEN 'Q2'
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9 THEN 'Q3'
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12 THEN 'Q4'
    END AS quarter_filter,
    COUNT(*) AS total_record_filter
    FROM 
        return_cleaned
    WHERE 
        return_reason_cleaned = 'Salah Input Sistem'
    GROUP BY 
        quarter_filter
)


SELECT
    CASE
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3 THEN 'Q1'
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6 THEN 'Q2'
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9 THEN 'Q3'
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12 THEN 'Q4'
    END AS quarter,
    qr.total_record_filter,
    COUNT(*) AS total_record,
    ROUND((NULLIF(qr.total_record_filter,0) * 1.0 / NULLIF(COUNT(*), 0))*100, 2) AS rate
FROM 
    return_cleaned rc
LEFT JOIN qr
    ON CASE
        WHEN EXTRACT(MONTH FROM rc.return_date) BETWEEN 1 AND 3 THEN 'Q1'
        WHEN EXTRACT(MONTH FROM rc.return_date) BETWEEN 4 AND 6 THEN 'Q2'
        WHEN EXTRACT(MONTH FROM rc.return_date) BETWEEN 7 AND 9 THEN 'Q3'
        WHEN EXTRACT(MONTH FROM rc.return_date) BETWEEN 10 AND 12 THEN 'Q4'
    END = qr.quarter_filter
GROUP BY 
    quarter,
    qr.total_record_filter
ORDER BY 
    quarter

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


quarter,total_record_filter,total_record,rate
Q1,2,8,25.00
Q2,10,15,66.67
Q3,6,14,42.86
Q4,10,22,45.45


**Findings**  
On a quarterly basis, Q2 recorded the highest proportion of Salah Input Sistem returns (66.67% of 15 return cases), 19.21 percentage points above the overall baseline (47.46%). However, since the data covers only a single year (2024), this Q2 spike cannot yet be confirmed as a recurring seasonal pattern — multi-year data would be needed to determine whether this is a repeating trend or a one-off occurrence.

In [17]:
%%sql

SELECT
    region_cleaned,
    COUNT (*) AS total_record
FROM
    return_cleaned
WHERE
    return_reason_cleaned = 'Salah Input Sistem'
GROUP BY
    region_cleaned
ORDER BY
    total_record DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
9 rows affected.


region_cleaned,total_record
Jakarta Barat,8
Jakarta Selatan,5
Bekasi,5
Jakarta Pusat,3
Bogor,2
Jakarta Timur,2
Depok,1
Jakarta Utara,1
Tangerang,1


In [18]:
%%sql

WITH rrc AS (
    SELECT
        region_cleaned,
        COUNT(*) AS total_record_filter
    FROM
        return_cleaned
    WHERE
        return_reason_cleaned = 'Salah Input Sistem'
    GROUP BY
        region_cleaned
)
SELECT
    rc.region_cleaned,
    rrc.total_record_filter,
    COUNT(*) AS total_record_all,
    ROUND((NULLIF(rrc.total_record_filter,0) * 1.0 / NULLIF(COUNT(*), 0))*100, 2) AS rate
FROM
    return_cleaned rc
LEFT JOIN rrc
    ON rc.region_cleaned = rrc.region_cleaned
GROUP BY
    rc.region_cleaned,
    rrc.total_record_filter
ORDER BY
    rrc.total_record_filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
9 rows affected.


region_cleaned,total_record_filter,total_record_all,rate
Jakarta Barat,8,13,61.54
Bekasi,5,10,50.00
Jakarta Selatan,5,10,50.00
Jakarta Pusat,3,3,100.00
Bogor,2,4,50.00
Jakarta Timur,2,6,33.33
Depok,1,3,33.33
Jakarta Utara,1,5,20.00
Tangerang,1,5,20.00


**Findings**  
Jakarta Barat recorded the highest proportion of Salah Input Sistem returns worth noting — 8 of 13 return cases (61.54%), about 14.08 percentage points above the overall baseline (47.46%). Other regions such as Jakarta Pusat (100%) and Bogor (50%) also showed high rates, but these cannot be considered significant given their very small sample sizes (n=3 and n=4), which makes them prone to small-sample bias.

In [21]:
%%sql

SELECT
    kategori,
    COUNT (*) AS total_record
FROM
    return_cleaned
WHERE
    return_reason_cleaned = 'Salah Input Sistem'
GROUP BY
   kategori
ORDER BY
    total_record DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


kategori,total_record
Diagnostik,9
Consumable,6
Bedah & Tindakan,5
Laboratorium,5
Perawatan & Rawat Inap,3


In [20]:
%%sql

WITH rk AS (
    SELECT
        kategori,
        COUNT (*) AS total_record
    FROM
        return_cleaned
    WHERE
        return_reason_cleaned = 'Salah Input Sistem'
    GROUP BY
        kategori
)


SELECT
    rc.kategori,
    rk.total_record AS total_record_filter,
    COUNT (*) AS total_record,
    ROUND((NULLIF(rk.total_record,0) * 1.0 / NULLIF(COUNT(*), 0))*100, 2) AS rate
FROM
    return_cleaned rc
LEFT JOIN rk
    ON rc.kategori = rk.kategori
GROUP BY
   rc.kategori,
    rk.total_record
ORDER BY
    total_record DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


kategori,total_record_filter,total_record,rate
Diagnostik,9,16,56.25
Consumable,6,13,46.15
Bedah & Tindakan,5,11,45.45
Perawatan & Rawat Inap,3,11,27.27
Laboratorium,5,8,62.50


**Findings**  
Diagnostik was the product category with the highest proportion of Salah Input Sistem returns worth noting — 9 of 16 return cases (56.25%), about 8.79 percentage points above the overall baseline (47.46%). Laboratorium recorded an even higher rate (62.50%), but with a smaller sample size (n=8), making it a less reliable signal compared to Diagnostik.

In [ ]:
%%sql



SELECT
    customer_type_cleaned,
    COUNT (*) AS total_record
FROM
    return_cleaned
WHERE
    return_reason_cleaned = 'Salah Input Sistem'
GROUP BY
   customer_type_cleaned
ORDER BY
    total_record DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
3 rows affected.


customer_type_cleaned,total_record
RS Swasta,11
Klinik,9
RS Pemerintah,8


In [37]:
%%sql

WITH ctc AS (
    SELECT
        customer_type_cleaned,
        COUNT (*) AS total_record
    FROM
        return_cleaned
    WHERE
        return_reason_cleaned = 'Salah Input Sistem'
    GROUP BY
        customer_type_cleaned
)

SELECT
    rc.customer_type_cleaned,
    ctc.total_record AS total_record_filter,
    COUNT (*) AS total_record,
    ROUND((NULLIF(ctc.total_record,0) * 1.0 / NULLIF(COUNT(*), 0))*100, 2) AS ratio
FROM
    return_cleaned rc
LEFT JOIN ctc
    ON rc.customer_type_cleaned = ctc.customer_type_cleaned
GROUP BY
   rc.customer_type_cleaned,
   ctc.total_record
ORDER BY
    total_record DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
3 rows affected.


customer_type_cleaned,total_record_filter,total_record,ratio
RS Swasta,11,27,40.74
Klinik,9,19,47.37
RS Pemerintah,8,13,61.54


**Findings**  
RS Pemerintah showed the highest rate of Salah Input Sistem returns among customer types (61.54% of 13 cases), 14.08 percentage points above the baseline. In contrast, RS Swasta — despite recording the highest total number of return cases overall (27 cases) — had the lowest rate (40.74%), 6.72 percentage points below baseline, indicating that Salah Input Sistem is relatively less prominent among RS Swasta's returns compared to other reasons. Klinik sat close to the baseline (47.37%).

In [ ]:
%%sql

SELECT
    region_cleaned,
    COUNT(*) FILTER (WHERE kategori = 'Bedah & Tindakan' ) AS Bedah_and_Tindakan,
    COUNT(*) FILTER (WHERE kategori = 'Diagnostik' ) AS Diagnostik,
    COUNT(*) FILTER (WHERE kategori = 'Consumable' ) AS Consumable,
    COUNT(*) FILTER (WHERE kategori = 'Laboratorium' ) AS Laboratorium,
    COUNT(*) FILTER (WHERE kategori = 'Perawatan & Rawat Inap' ) AS Perawatan_and_Rawat_Inap,
    COUNT(*) AS total_records
FROM
    return_cleaned
WHERE
    return_reason_cleaned = 'Salah Input Sistem'
GROUP BY
    region_cleaned
ORDER BY
    region_cleaned

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
9 rows affected.


region_cleaned,bedah_and_tindakan,diagnostik,consumable,laboratorium,perawatan_and_rawat_inap,total_records
Bekasi,1,1,0,3,0,5
Bogor,0,0,1,1,0,2
Depok,0,1,0,0,0,1
Jakarta Barat,1,4,1,1,1,8
Jakarta Pusat,1,1,1,0,0,3
Jakarta Selatan,0,2,2,0,1,5
Jakarta Timur,0,0,1,0,1,2
Jakarta Utara,1,0,0,0,0,1
Tangerang,1,0,0,0,0,1


**Findings**  
Jakarta Barat × Diagnostik recorded 4 cases, compared to an expected 2.57 cases if region and category were unrelated — 1.43 cases higher (about 1.56x the expected count). However, this combination accounts for only 14.28% (4 of 28) of total Salah Input Sistem cases, so despite the deviation from expectation, its overall impact on the broader problem remains small. With small sample sizes on both sides (Jakarta Barat n=8, Diagnostik n=9, both below the reliability threshold), this finding is indicative only and not yet strong enough to be considered a confirmed pattern.

In [4]:
%%sql

SELECT
    customer_type_cleaned,
    COUNT(*) FILTER (WHERE kategori = 'Bedah & Tindakan' ) AS Bedah_and_Tindakan,
    COUNT(*) FILTER (WHERE kategori = 'Diagnostik' ) AS Diagnostik,
    COUNT(*) FILTER (WHERE kategori = 'Consumable' ) AS Consumable,
    COUNT(*) FILTER (WHERE kategori = 'Laboratorium' ) AS Laboratorium,
    COUNT(*) FILTER (WHERE kategori = 'Perawatan & Rawat Inap' ) AS Perawatan_and_Rawat_Inap,
    COUNT(*) AS total_records
FROM
    return_cleaned
WHERE
    return_reason_cleaned = 'Salah Input Sistem'
GROUP BY
    customer_type_cleaned
ORDER BY
    customer_type_cleaned

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
3 rows affected.


customer_type_cleaned,bedah_and_tindakan,diagnostik,consumable,laboratorium,perawatan_and_rawat_inap,total_records
Klinik,0,4,3,2,0,9
RS Pemerintah,2,1,2,2,1,8
RS Swasta,3,4,1,1,2,11


**Findings**  
RS Pemerintah × Diagnostik recorded 1 case, compared to an expected 2.57 cases if customer type and category were unrelated — 1.57 cases lower (about 0.4x the expected count). This combination accounts for only 3.57% (1 of 28 cases) of total Salah Input Sistem cases, so despite being well below expectation, its impact on the overall problem remains small. With small sample sizes on both sides (RS Pemerintah=8, Diagnostik=9, both below the reliability threshold), this finding is indicative only and not yet strong enough to be considered a confirmed pattern.

In [36]:
%%sql

SELECT
    region_cleaned,
    COUNT(*) FILTER (WHERE customer_type_cleaned = 'RS Swasta' ) AS RS_Swasta,
    COUNT(*) FILTER (WHERE customer_type_cleaned = 'Klinik' ) AS Klinik,
    COUNT(*) FILTER (WHERE customer_type_cleaned = 'RS Pemerintah' ) AS RS_Pemerintah,
    COUNT(*) AS total_records
FROM
    return_cleaned
WHERE
    return_reason_cleaned = 'Salah Input Sistem'
GROUP BY
    region_cleaned
ORDER BY
    region_cleaned

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
9 rows affected.


region_cleaned,rs_swasta,klinik,rs_pemerintah,total_records
Bekasi,1,2,2,5
Bogor,0,2,0,2
Depok,1,0,0,1
Jakarta Barat,4,1,3,8
Jakarta Pusat,1,2,0,3
Jakarta Selatan,2,1,2,5
Jakarta Timur,0,1,1,2
Jakarta Utara,1,0,0,1
Tangerang,1,0,0,1


**Findings**  
The Jakarta Barat × RS Pemerintah combination recorded 3 cases, compared to an expected 2.29 cases if region and customer type were unrelated — 0.7 cases higher (about 1.3x the expected count). This combination accounts for only 10.7% (3 of 28 cases) of total Salah Input Sistem cases, so despite the deviation from expectation, its impact on the overall problem remains small. With small sample sizes on both sides (Jakarta Barat=8, RS Pemerintah=8, both below the reliability threshold), this finding is indicative only and not yet strong enough to be considered a confirmed pattern.

**Insight**  
Salah Input Sistem was the leading return reason (28 of 59 cases, 47.46%). Jakarta Barat showed a notable concentration of Diagnostik-category cases — 4 cases compared to an expected 2.57 if the two dimensions were unrelated (about 14.28% of total Salah Input Sistem cases). A similar, weaker signal appeared for Jakarta Barat combined with RS Pemerintah (3 cases vs. an expected 2.29). RS Pemerintah, however, is likely not the primary driver here — without Jakarta Barat, its link to Diagnostik disappears (1 case vs. an expected 2.57), suggesting Jakarta Barat itself is the main point of concentration, not RS Pemerintah. The exact cause of this concentration cannot yet be confirmed — it may relate to input process, product variant complexity, or admin workload — and requires further investigation. It's worth noting that all these combinations have small sample sizes on both sides (below the reliability threshold), so these findings remain indicative. Regardless of the confirmed cause, this Jakarta Barat–Diagnostik concentration still implies potential losses from reprocessing, staff time spent handling returns, and the risk of returned goods becoming unsellable — making this point worth investigating and addressing first, ahead of other points that show no comparably strong pattern.

In [5]:
%%sql

SELECT
    kategori,
    COUNT(*) AS total_record
FROM
    return_cleaned
GROUP BY
    kategori
ORDER BY
    total_record DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


kategori,total_record
Diagnostik,16
Consumable,13
Bedah & Tindakan,11
Perawatan & Rawat Inap,11
Laboratorium,8


In [7]:
%%sql

WITH
    kategori_total AS (
        SELECT
            COUNT(*) AS total_records
        FROM
            return_cleaned
    )

SELECT
    kategori,
    COUNT(*) AS total_records_filter,
    kt.total_records,
    ROUND((COUNT(*)::numeric / NULLIF(kt.total_records,0)::numeric) * 100,2) AS rate
FROM
    return_cleaned rc
CROSS JOIN 
    kategori_total kt
GROUP BY
    kategori,
    kt.total_records
ORDER BY
    total_records_filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


kategori,total_records_filter,total_records,rate
Diagnostik,16,59,27.12
Consumable,13,59,22.03
Perawatan & Rawat Inap,11,59,18.64
Bedah & Tindakan,11,59,18.64
Laboratorium,8,59,13.56


**Findings** 
Diagnostik had the highest number of return cases (16 of 59, 27.12%), followed by Consumable (13 cases, 22.03%). Perawatan & Rawat Inap and Bedah & Tindakan recorded the same count (11 cases each, 18.64%), while Laboratorium had the fewest return cases (8 cases, 13.56%).

In [8]:
%%sql

SELECT
    sku_name,
    COUNT(*) AS total_record
FROM
    return_cleaned
WHERE
    kategori = 'Diagnostik'
GROUP BY
    sku_name
ORDER BY
    total_record DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
10 rows affected.


sku_name,total_record
Urine Stick 10 Parameter,3
Pulse Oximeter,2
Glukometer Set,2
Rapid Test HbsAg,2
Tensimeter Manual Dewasa,2
Tensimeter Digital Lengan,1
Stethoscope Dewasa,1
Stethoscope Anak,1
Penlight Diagnostik,1
Timbangan Bayi Digital,1


In [6]:
%%sql

WITH
    sku_total AS (
        SELECT
            COUNT(*) AS total_loss
        FROM
            return_cleaned
        WHERE
            kategori = 'Diagnostik'

    ),
    sku_filter AS (
        SELECT
            sku_name,
            COUNT(*) AS total_loss_filter
        FROM
            return_cleaned
        WHERE
            kategori = 'Diagnostik'
        GROUP BY
            sku_name
    )

SELECT
    sf.sku_name,
    sf.total_loss_filter,
    st.total_loss,
    ROUND((sf.total_loss_filter::numeric / NULLIF(st.total_loss,0)::numeric) * 100,2) AS rate
FROM
    sku_filter sf
CROSS JOIN
    sku_total st
GROUP BY
    sf.sku_name,
    sf.total_loss_filter,
    st.total_loss
ORDER BY
    sf.total_loss_filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
10 rows affected.


sku_name,total_loss_filter,total_loss,rate
Urine Stick 10 Parameter,3,16,18.75
Glukometer Set,2,16,12.50
Pulse Oximeter,2,16,12.50
Rapid Test HbsAg,2,16,12.50
Tensimeter Manual Dewasa,2,16,12.50
Penlight Diagnostik,1,16,6.25
Stethoscope Anak,1,16,6.25
Stethoscope Dewasa,1,16,6.25
Tensimeter Digital Lengan,1,16,6.25
Timbangan Bayi Digital,1,16,6.25


**Findings**  
Urine Stick 10 Parameter had the most returns within the Diagnostik category (3 of 16 cases, 18.75%), followed by Glukometer Set, Pulse Oximeter, Rapid Test HbsAg, and Tensimeter Manual Dewasa, each with 2 cases (12.50%). Given the very small sample size per product (a maximum of 3 cases out of 16) and a relatively even spread (6.25%–18.75%, with no product standing out from the rest), this product-level breakdown isn't reliable enough to serve as a basis for further analysis. As a result, the analysis stops at this level — the search for causes for BQ2 continues at the Diagnostik category level instead of any single SKU.

In [11]:
%%sql

SELECT
    CASE
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3 THEN 'Q1'
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6 THEN 'Q2'
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9 THEN 'Q3'
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12 THEN 'Q4'
    END AS quarter,
    COUNT(*) AS total_record
FROM
    return_cleaned
WHERE
    kategori = 'Diagnostik'
GROUP BY
    quarter
ORDER BY
    total_record DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


quarter,total_record
Q4,5
Q3,4
Q2,4
Q1,3


In [10]:
%%sql

WITH 
month AS (
    SELECT
        CASE
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3 THEN 'Q1'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6 THEN 'Q2'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9 THEN 'Q3'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12 THEN 'Q4'
        END AS quarter,
        COUNT(value_return) AS total_records
    FROM
        return_cleaned
    GROUP BY
        quarter
),
month_filter AS (
    SELECT
            CASE
                WHEN EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3 THEN 'Q1'
                WHEN EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6 THEN 'Q2'
                WHEN EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9 THEN 'Q3'
                WHEN EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12 THEN 'Q4'
            END AS quarter,
            COUNT(value_return) AS total_records_filter
    FROM
        return_cleaned
    WHERE
        kategori = 'Diagnostik'
    GROUP BY
        quarter
)

SELECT
    m.quarter,
    mf.total_records_filter AS total_records_filter,
    m.total_records AS total_records_all,
    ROUND( (COALESCE(mf.total_records_filter,0) * 1.0 / NULLIF(m.total_records,0) * 100)::numeric,2) AS ratio

FROM
    month m
LEFT JOIN
    month_filter mf 
        ON m.quarter = mf.quarter
ORDER BY
    m.quarter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


quarter,total_records_filter,total_records_all,ratio
Q4,5,22,22.73
Q3,4,14,28.57
Q2,4,15,26.67
Q1,3,8,37.50


**Findings**  
Q1 showed the highest Diagnostik-category rate among all 2024 quarters (37.50% of 8 cases), 10.38 percentage points above the baseline (27.12%). However, given Q1's small sample size (n=8, below the reliability threshold), this finding is indicative and should be read with caution. In contrast, Q4 recorded the lowest rate (22.73%) despite having the highest overall return volume (22 cases across all categories) — suggesting that Q4's high case count is driven by overall transaction volume rather than a particular prominence of the Diagnostik category in that quarter.

In [12]:
%%sql

SELECT
    return_reason_cleaned,
    COUNT(*) AS total_records
FROM
    return_cleaned
WHERE
    kategori = 'Diagnostik'
GROUP BY
    return_reason_cleaned
ORDER BY
    total_records DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


return_reason_cleaned,total_records
Salah Input Sistem,9
Salah Pesan Customer,4
Barang Rusak,2
Kadaluarsa,1


In [20]:
%%sql

WITH kategori_reason AS
(
    SELECT
        return_reason_cleaned,
        COUNT(*) AS total_records_filter
    FROM
        return_cleaned
    WHERE
        kategori = 'Diagnostik'
    GROUP BY
        return_reason_cleaned
)

SELECT
    rc.return_reason_cleaned,
    ks.total_records_filter,
    COUNT(*) AS total_records,
    ROUND((ks.total_records_filter::numeric / NULLIF(COUNT(*),0)::numeric) * 100,2) AS ratio
FROM
    return_cleaned rc
LEFT JOIN 
    kategori_reason ks
        ON rc.return_reason_cleaned = ks.return_reason_cleaned
GROUP BY
    rc.return_reason_cleaned,
    ks.total_records_filter
ORDER BY
    ks.total_records_filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


return_reason_cleaned,total_records_filter,total_records,ratio
Salah Input Sistem,9,28,32.14
Salah Pesan Customer,4,20,20.00
Barang Rusak,2,7,28.57
Kadaluarsa,1,4,25.00


In [23]:
%%sql

SELECT
    customer_type_cleaned,
    COUNT(*) AS total_records
FROM
    return_cleaned
WHERE
    kategori = 'Diagnostik'
GROUP BY
    customer_type_cleaned
ORDER BY
    total_records DESC
    

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
3 rows affected.


customer_type_cleaned,total_records
Klinik,7
RS Swasta,6
RS Pemerintah,3


In [28]:
%%sql

WITH kategori_customer AS
(
    SELECT
        customer_type_cleaned,
        COUNT(*) AS total_records_filter
    FROM
        return_cleaned
    WHERE
        kategori = 'Diagnostik'
    GROUP BY
        customer_type_cleaned
)

SELECT
    rc.customer_type_cleaned,
    kc.total_records_filter,
    COUNT(*) AS total_records,
    ROUND((kc.total_records_filter::numeric / NULLIF(COUNT(*),0)::numeric) * 100,2) AS ratio
FROM
    return_cleaned rc
LEFT JOIN
    kategori_customer kc
        ON rc.customer_type_cleaned = kc.customer_type_cleaned
GROUP BY
    rc.customer_type_cleaned,
    kc.total_records_filter
ORDER BY
    kc.total_records_filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
3 rows affected.


customer_type_cleaned,total_records_filter,total_records,ratio
Klinik,7,19,36.84
RS Swasta,6,27,22.22
RS Pemerintah,3,13,23.08


In [30]:
%%sql

SELECT
    region_cleaned,
    COUNT(*) AS total_records
FROM
    return_cleaned
WHERE
    kategori = 'Diagnostik'
GROUP BY
    region_cleaned
ORDER BY
    total_records DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
7 rows affected.


region_cleaned,total_records
Jakarta Barat,7
Jakarta Selatan,3
Bekasi,2
Tangerang,1
Jakarta Pusat,1
Depok,1
Jakarta Utara,1


In [37]:
%%sql

WITH kategori_region AS
(
    SELECT
        region_cleaned,
        COUNT(*) AS total_records_filter
    FROM
        return_cleaned
    WHERE
        kategori = 'Diagnostik'
    GROUP BY
        region_cleaned
)

SELECT
    rc.region_cleaned,
    kg.total_records_filter,
    COUNT(*) AS total_filter,
    ROUND((kg.total_records_filter::numeric / NULLIF(COUNT(*),0)::numeric) * 100,2) AS ratio
FROM
    return_cleaned rc
LEFT JOIN
    kategori_region kg
        ON rc.region_cleaned = kg.region_cleaned
GROUP BY
    rc.region_cleaned,
    kg.total_records_filter
ORDER BY
    kg.total_records_filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
9 rows affected.


region_cleaned,total_records_filter,total_filter,ratio
Bogor,None,4,None
Jakarta Timur,None,6,None
Jakarta Barat,7,13,53.85
Jakarta Selatan,3,10,30.00
Bekasi,2,10,20.00
Depok,1,3,33.33
Jakarta Pusat,1,3,33.33
Jakarta Utara,1,5,20.00
Tangerang,1,5,20.00


**Insight**  


**Conclusion:** 
 

# 2. Financial Impact

In [64]:
%%sql

SELECT
    SUM(value_return) AS total_loss,
    AVG(value_return) AS avg
FROM
    return_cleaned


 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


total_loss,avg
171506500.0,2906889.8305084747


In [65]:
%%sql

SELECT
    kategori,
    SUM(value_return) AS total_loss,
    COUNT(value_return) AS total_record
FROM
    return_cleaned
GROUP BY
    kategori
ORDER BY
    total_loss DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


kategori,total_loss,total_record
Diagnostik,66811000.0,16
Bedah & Tindakan,53336500.0,11
Perawatan & Rawat Inap,22706500.0,11
Consumable,21474000.0,13
Laboratorium,7178500.0,8


In [ ]:
%%sql

WITH
    kategori_total AS (
        SELECT
            SUM(value_return) AS total_loss
        FROM
            return_cleaned
    )

SELECT
    kategori,
    SUM(value_return) AS total_loss_filter,
    kt.total_loss,
    ROUND((SUM(value_return)::numeric / NULLIF(kt.total_loss,0)::numeric) * 100,2) AS ratio,
    COUNT(value_return) AS total_record_filter,
    AVG(value_return) AS average_loss_filter
FROM
    return_cleaned rc
CROSS JOIN 
    kategori_total kt
GROUP BY
    kategori,
    kt.total_loss
ORDER BY
    total_loss_filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


kategori,total_loss_filter,total_loss,ratio,total_record_filter,average_loss_filter
Diagnostik,66811000.0,171506500.0,38.96,16,4175687.5
Bedah & Tindakan,53336500.0,171506500.0,31.10,11,4848772.7272727275
Perawatan & Rawat Inap,22706500.0,171506500.0,13.24,11,2064227.2727272727
Consumable,21474000.0,171506500.0,12.52,13,1651846.1538461538
Laboratorium,7178500.0,171506500.0,4.19,8,897312.5


In [5]:
%%sql

SELECT
    sku_name,
    SUM(value_return) AS total_loss,
    COUNT(value_return) AS total_record
FROM
    return_cleaned
WHERE
    kategori = 'Diagnostik'
GROUP BY
    sku_name
ORDER BY
    total_loss DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
10 rows affected.


sku_name,total_loss,total_record
Rapid Test HbsAg,24174000.0,2
Urine Stick 10 Parameter,19189000.0,3
Glukometer Set,7980000.0,2
Pulse Oximeter,5077000.0,2
Stethoscope Anak,2761000.0,1
Tensimeter Manual Dewasa,2377000.0,2
Stethoscope Dewasa,1834000.0,1
Timbangan Bayi Digital,1587000.0,1
Tensimeter Digital Lengan,1468000.0,1
Penlight Diagnostik,364000.0,1


In [20]:
%%sql

SELECT
        CASE
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3 THEN 'Q1'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6 THEN 'Q2'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9 THEN 'Q3'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12 THEN 'Q4'
        END AS quarter,
        SUM(value_return) AS total_loss
FROM
    return_cleaned
GROUP BY
    quarter

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


quarter,total_loss
Q3,34531500.0
Q1,30449500.0
Q2,69647000.0
Q4,36878500.0


In [10]:
%%sql

SELECT
        CASE
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3 THEN 'Q1'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6 THEN 'Q2'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9 THEN 'Q3'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12 THEN 'Q4'
        END AS quarter,
        SUM(value_return) AS total_loss
FROM
    return_cleaned
WHERE
    kategori = 'Diagnostik'
GROUP BY
    quarter

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


quarter,total_loss
Q1,3510000.0
Q3,25298000.0
Q4,4608000.0
Q2,33395000.0


In [52]:
%%sql

WITH 
month AS (
    SELECT
        CASE
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3 THEN 'Q1'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6 THEN 'Q2'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9 THEN 'Q3'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12 THEN 'Q4'
        END AS quarter,
        SUM(value_return) AS total_loss
    FROM
        return_cleaned
    GROUP BY
        quarter
),
month_filter AS (
    SELECT
            CASE
                WHEN EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3 THEN 'Q1'
                WHEN EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6 THEN 'Q2'
                WHEN EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9 THEN 'Q3'
                WHEN EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12 THEN 'Q4'
            END AS quarter,
            SUM(value_return) AS total_loss
    FROM
        return_cleaned
    WHERE
        kategori = 'Diagnostik'
    GROUP BY
        quarter
)

SELECT
    m.quarter,
    mf.total_loss AS total_loss,
    m.total_loss AS total_loss_all,
    ROUND( (COALESCE(mf.total_loss,0) * 1.0 / NULLIF(m.total_loss,0) * 100)::numeric,2) AS ratio

FROM
    month m
LEFT JOIN
    month_filter mf 
        ON m.quarter = mf.quarter
ORDER BY
    m.quarter

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


quarter,total_loss,total_loss_all,ratio
Q1,3510000.0,30449500.0,11.53
Q2,33395000.0,69647000.0,47.95
Q3,25298000.0,34531500.0,73.26
Q4,4608000.0,36878500.0,12.50


In [32]:
%%sql

SELECT
    return_reason_cleaned,
    SUM(value_return) AS total_loss,
    COUNT(return_reason_cleaned) AS frequency
FROM
    return_cleaned
WHERE
    kategori = 'Diagnostik'
GROUP BY
    return_reason_cleaned
ORDER BY
    total_loss DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


return_reason_cleaned,total_loss,frequency
Salah Input Sistem,26842000.0,9
Salah Pesan Customer,19761000.0,4
Barang Rusak,17447000.0,2
Kadaluarsa,2761000.0,1


In [35]:
%%sql

SELECT
    return_reason_cleaned,
    SUM(value_return) AS total_loss,
    COUNT(return_reason_cleaned) AS frequency
FROM
    return_cleaned
GROUP BY
    return_reason_cleaned
ORDER BY
    total_loss DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


return_reason_cleaned,total_loss,frequency
Salah Pesan Customer,70452500.0,20
Salah Input Sistem,66634500.0,28
Barang Rusak,29144000.0,7
Kadaluarsa,5275500.0,4


In [52]:
%%sql

WITH kategori_reason AS (
    SELECT
        return_reason_cleaned,
        SUM(value_return) AS total_loss_filter
    FROM
        return_cleaned
    WHERE
        kategori = 'Diagnostik'
    GROUP BY
        return_reason_cleaned       
)

SELECT
    rc.return_reason_cleaned,
    ks.total_loss_filter,
    SUM(rc.value_return) AS total_loss,
    ROUND( (COALESCE(ks.total_loss_filter,0) * 1.0 / NULLIF(SUM(rc.value_return),0)* 100)::numeric,2) AS ratio
FROM
    return_cleaned rc
LEFT JOIN
    kategori_reason ks
    ON rc.return_reason_cleaned = ks.return_reason_cleaned
GROUP BY
    rc.return_reason_cleaned,
    ks.total_loss_filter
ORDER BY
    ks.total_loss_filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


return_reason_cleaned,total_loss_filter,total_loss,ratio
Salah Input Sistem,26842000.0,66634500.0,40.28
Salah Pesan Customer,19761000.0,70452500.0,28.05
Barang Rusak,17447000.0,29144000.0,59.86
Kadaluarsa,2761000.0,5275500.0,52.34


In [55]:
%%sql

SELECT
    customer_type_cleaned,
    SUM(value_return) AS total_loss,
    COUNT(value_return) AS frequency
FROM
    return_cleaned
WHERE
    kategori = 'Diagnostik'
GROUP BY
    customer_type_cleaned
ORDER BY
    total_loss DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
3 rows affected.


customer_type_cleaned,total_loss,frequency
Klinik,29225000.0,7
RS Pemerintah,19281000.0,3
RS Swasta,18305000.0,6


In [56]:
%%sql

SELECT
    customer_type_cleaned,
    SUM(value_return) AS total_loss,
    COUNT(value_return) AS frequency
FROM
    return_cleaned
GROUP BY
    customer_type_cleaned
ORDER BY
    total_loss DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
3 rows affected.


customer_type_cleaned,total_loss,frequency
RS Swasta,76510000.0,27
Klinik,52240500.0,19
RS Pemerintah,42756000.0,13


In [58]:
%%sql

WITH kategori_customer AS (
    SELECT
        customer_type_cleaned,
        SUM(value_return) AS total_loss_filter
    FROM
        return_cleaned
    WHERE
        kategori = 'Diagnostik'
    GROUP BY
        customer_type_cleaned       
)

SELECT
    rc.customer_type_cleaned,
    kc.total_loss_filter,
    SUM(rc.value_return) AS total_loss,
    ROUND( (COALESCE(kc.total_loss_filter,0) * 1.0 / NULLIF(SUM(rc.value_return),0)* 100)::numeric,2) AS ratio
FROM
    return_cleaned rc
LEFT JOIN
    kategori_customer kc
    ON rc.customer_type_cleaned = kc.customer_type_cleaned
GROUP BY
    rc.customer_type_cleaned,
    kc.total_loss_filter
ORDER BY
    kc.total_loss_filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
3 rows affected.


customer_type_cleaned,total_loss_filter,total_loss,ratio
Klinik,29225000.0,52240500.0,55.94
RS Pemerintah,19281000.0,42756000.0,45.10
RS Swasta,18305000.0,76510000.0,23.92


In [59]:
%%sql

SELECT *
FROM
    return_cleaned
LIMIT 3

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
3 rows affected.


retur_id,invoice_id,sku_id,customer_id,return_date,sku_name,kategori,value_return,return_reason_cleaned,customer_type_cleaned,region_cleaned,upper_bound
RT00042,010029,D001,RSP053,2024-03-02,Tensimeter Digital Lengan,Diagnostik,1468000.0,Barang Rusak,RS Pemerintah,Jakarta Utara,4107875.0
RT00046,010092,B002,RSS044,2024-05-14,Gunting Bedah Bengkok 14cm,Bedah & Tindakan,417500.0,Salah Pesan Customer,RS Swasta,Jakarta Selatan,4107875.0
RT00021,009825,L002,RSS056,2024-11-22,Tabung Vacutainer Plain,Laboratorium,478000.0,Salah Input Sistem,RS Swasta,Bekasi,4107875.0


In [61]:
%%sql

SELECT
    region_cleaned,
    SUM(value_return) AS total_loss,
    COUNT(value_return) AS frequency
FROM
    return_cleaned
GROUP BY
    region_cleaned
ORDER BY
    total_loss DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
9 rows affected.


region_cleaned,total_loss,frequency
Bekasi,36964500.0,10
Jakarta Barat,33920000.0,13
Bogor,21657000.0,4
Tangerang,20677000.0,5
Jakarta Utara,19318500.0,5
Jakarta Pusat,19224000.0,3
Jakarta Selatan,11695500.0,10
Jakarta Timur,4650500.0,6
Depok,3399500.0,3


In [62]:
%%sql

WITH kategori_region AS (
    SELECT
        region_cleaned,
        SUM(value_return) AS total_loss_filter
    FROM
        return_cleaned
    WHERE
        kategori = 'Diagnostik'
    GROUP BY
        region_cleaned       
)

SELECT
    rc.region_cleaned,
    kr.total_loss_filter,
    SUM(rc.value_return) AS total_loss,
    ROUND( (COALESCE(kr.total_loss_filter,0) * 1.0 / NULLIF(SUM(rc.value_return),0)* 100)::numeric,2) AS ratio
FROM
    return_cleaned rc
LEFT JOIN
    kategori_region kr
    ON rc.region_cleaned = kr.region_cleaned
GROUP BY
    rc.region_cleaned,
    kr.total_loss_filter
ORDER BY
    kr.total_loss_filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
9 rows affected.


region_cleaned,total_loss_filter,total_loss,ratio
Bogor,None,21657000.0,0.00
Jakarta Timur,None,4650500.0,0.00
Jakarta Barat,19953000.0,33920000.0,58.82
Bekasi,18734000.0,36964500.0,50.68
Tangerang,15979000.0,20677000.0,77.28
Jakarta Selatan,7182000.0,11695500.0,61.41
Depok,1908000.0,3399500.0,56.13
Jakarta Pusat,1587000.0,19224000.0,8.26
Jakarta Utara,1468000.0,19318500.0,7.60


**Conclusion:** 
 

# 3. Who & Where

In [39]:
%%sql

SELECT
    customer_type_cleaned,
    COUNT(*) AS total_records
FROM
    return_cleaned
GROUP BY
    customer_type_cleaned
ORDER BY
    total_records DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
3 rows affected.


customer_type_cleaned,total_records
RS Swasta,27
Klinik,19
RS Pemerintah,13


In [50]:
%%sql

WITH customer_type AS
(
    SELECT
        COUNT(*) AS total_records
    FROM
        return_cleaned
)

SELECT
    rc.customer_type_cleaned,
    COUNT(*) AS total_records__filter,
    ct.total_records AS total_records,
    ROUND((COUNT(*)::numeric / NULLIF(ct.total_records,0)::numeric) * 100,2) AS ratio
FROM
    return_cleaned rc
CROSS JOIN 
    customer_type ct
GROUP BY
    rc.customer_type_cleaned,
    ct.total_records
ORDER BY
    total_records__filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
3 rows affected.


customer_type_cleaned,total_records__filter,total_records,ratio
RS Swasta,27,59,45.76
Klinik,19,59,32.20
RS Pemerintah,13,59,22.03


In [51]:
%%sql

SELECT
    CASE
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3 THEN 'Q1'
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6 THEN 'Q2'
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9 THEN 'Q3'
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12 THEN 'Q4'
    END AS quarter,
    COUNT(*) AS total_records
FROM
    return_cleaned
WHERE
    customer_type_cleaned ='RS Swasta'
GROUP BY
    quarter
ORDER BY
    total_records DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


quarter,total_records
Q4,10
Q2,8
Q3,8
Q1,1


In [59]:
%%sql

WITH 
ct_month AS (
    SELECT
        CASE
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3 THEN 'Q1'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6 THEN 'Q2'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9 THEN 'Q3'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12 THEN 'Q4'
        END AS quarter,
        COUNT(*) AS total_records
    FROM
        return_cleaned
    GROUP BY
        quarter
),
ct_month_filter AS (
    SELECT
            CASE
                WHEN EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3 THEN 'Q1'
                WHEN EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6 THEN 'Q2'
                WHEN EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9 THEN 'Q3'
                WHEN EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12 THEN 'Q4'
            END AS quarter,
            COUNT(*) AS total_records_filter
    FROM
        return_cleaned
    WHERE
        customer_type_cleaned ='RS Swasta'
    GROUP BY
        quarter
)

SELECT
    ctm.quarter,
    ctmf.total_records_filter,
    ctm.total_records,
    ROUND((ctmf.total_records_filter::numeric / NULLIF(ctm.total_records,0)::numeric) * 100,2) AS ratio

FROM
    ct_month ctm
LEFT JOIN
    ct_month_filter ctmf 
        ON ctm.quarter = ctmf.quarter
ORDER BY
    ctmf.total_records_filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


quarter,total_records_filter,total_records,ratio
Q4,10,22,45.45
Q3,8,14,57.14
Q2,8,15,53.33
Q1,1,8,12.50


In [60]:
%%sql

SELECT
    kategori,
    COUNT(*) AS total_records
FROM
    return_cleaned
WHERE
    customer_type_cleaned ='RS Swasta'
GROUP BY
    kategori
ORDER BY
    total_records DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


kategori,total_records
Bedah & Tindakan,7
Perawatan & Rawat Inap,7
Diagnostik,6
Laboratorium,4
Consumable,3


In [65]:
%%sql

SELECT
    return_reason_cleaned,
    COUNT(*) AS total_records
FROM
    return_cleaned
WHERE
    customer_type_cleaned ='RS Swasta'
GROUP BY
    return_reason_cleaned
ORDER BY
    total_records DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


return_reason_cleaned,total_records
Salah Input Sistem,11
Salah Pesan Customer,10
Barang Rusak,4
Kadaluarsa,2


In [66]:
%%sql

WITH ct_kategori_filter AS
(
    SELECT
        return_reason_cleaned,
        COUNT(*) AS total_records
    FROM
        return_cleaned
    WHERE
        customer_type_cleaned ='RS Swasta'
    GROUP BY
        return_reason_cleaned
)

SELECT
    rc.return_reason_cleaned,
    ctkf.total_records AS total_records_filter,
    COUNT(*) AS total_records,
    ROUND((ctkf.total_records::numeric / NULLIF(COUNT(*),0)::numeric) * 100,2) AS ratio
FROM
    return_cleaned rc
LEFT JOIN
    ct_kategori_filter ctkf
        ON rc.return_reason_cleaned = ctkf.return_reason_cleaned
GROUP BY
    rc.return_reason_cleaned,
    total_records_filter
ORDER BY
    total_records_filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


return_reason_cleaned,total_records_filter,total_records,ratio
Salah Input Sistem,11,28,39.29
Salah Pesan Customer,10,20,50.00
Barang Rusak,4,7,57.14
Kadaluarsa,2,4,50.00


In [69]:
%%sql

SELECT
    region_cleaned,
    COUNT(*) AS total_records
FROM
    return_cleaned
WHERE
    customer_type_cleaned ='RS Swasta'
GROUP BY
    region_cleaned
ORDER BY
    total_records DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
8 rows affected.


region_cleaned,total_records
Jakarta Barat,6
Jakarta Selatan,5
Tangerang,4
Bekasi,4
Jakarta Utara,3
Depok,2
Jakarta Timur,2
Jakarta Pusat,1


In [68]:
%%sql

SELECT *
FROM
    return_cleaned
LIMIT 3

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
3 rows affected.


retur_id,invoice_id,sku_id,customer_id,return_date,sku_name,kategori,value_return,return_reason_cleaned,customer_type_cleaned,region_cleaned,upper_bound
RT00042,010029,D001,RSP053,2024-03-02,Tensimeter Digital Lengan,Diagnostik,1468000.0,Barang Rusak,RS Pemerintah,Jakarta Utara,4107875.0
RT00046,010092,B002,RSS044,2024-05-14,Gunting Bedah Bengkok 14cm,Bedah & Tindakan,417500.0,Salah Pesan Customer,RS Swasta,Jakarta Selatan,4107875.0
RT00021,009825,L002,RSS056,2024-11-22,Tabung Vacutainer Plain,Laboratorium,478000.0,Salah Input Sistem,RS Swasta,Bekasi,4107875.0


In [70]:
%%sql

SELECT
    region_cleaned,
    COUNT(*) AS total_records
FROM
    return_cleaned
GROUP BY
    region_cleaned
ORDER BY
    total_records DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
9 rows affected.


region_cleaned,total_records
Jakarta Barat,13
Bekasi,10
Jakarta Selatan,10
Jakarta Timur,6
Tangerang,5
Jakarta Utara,5
Bogor,4
Jakarta Pusat,3
Depok,3


In [ ]:
%%sql

WITH region AS
(
    SELECT
        COUNT(*) AS total_records
    FROM
        return_cleaned
)

SELECT
    rc.region_cleaned,
    COUNT(*) AS total_records__filter,
    g.total_records AS total_records,
    ROUND((COUNT(*)::numeric / NULLIF(g.total_records,0)::numeric) * 100,2) AS rate
    return_cleaned rc
CROSS JOIN 
    region g
GROUP BY
    rc.region_cleaned,
    g.total_records
ORDER BY
    total_records__filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
9 rows affected.


region_cleaned,total_records__filter,total_records,ratio
Jakarta Barat,13,59,22.03
Bekasi,10,59,16.95
Jakarta Selatan,10,59,16.95
Jakarta Timur,6,59,10.17
Jakarta Utara,5,59,8.47
Tangerang,5,59,8.47
Bogor,4,59,6.78
Depok,3,59,5.08
Jakarta Pusat,3,59,5.08


**Conclusion:** 
 

# 4. Trend Over Time

**Conclusion:** 
 